Fix dependency copy for as_doc

svlandeg opened the pull request:
The `span.as_doc` functionality seems to have some issues (cf. Issue #3962 and Issue #3669) which I think are related to the dependency graph being cut into pieces. Some tokens may still refer to other tokens outside the `span`, thus resulting in an out of bounds error when that token is being accessed.

## Description
In `doc.from_array` I have added code that checks the offset of the head for each token, and does not copy the value unless that token is inside the doc : `if (i+value) in range(self.length)`. The token head defaults to 0 (i.e. itself) if the value is not copied/set.

This fix feels a little hacky, so happy to get feedback. Also currently the actual dependency link (e.g. `ccomp`) is being copied even if the head isn't, but I think we need that to ensure other cases keep working where a token refers to itself as head.

### Types of change
bug fix

## Checklist
- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

honnibal commented:
Hmm, yes I see the problem. This is a pretty nasty case :(.

It does make sense to handle these as fragments, but the problem is we'll be getting the fragments with `doc.sents`, which feels pretty undesirable.

Perhaps we could do the following to reconcile the tree:

* Find all the words whose head is outside the span.
* For each of those words, look for an ancestor that's within the span (using the `token.ancestors` attribute.) Attach to the lowest within-span ancestor, using the label `dep`.
* If there is no within-span ancestor, attach to the previous `ROOT` token using the label `dep`.

The `dep` label is kind of the miscellaneous one, used for tree-shapes that the PSG-to-dependencies converter couldn't classify into a more specific construction. So it's the best one to use for these cases where we're just stitching things together into a tree, even if it doesn't make good linguistic sense.

ines commented on spacy/tests/regression/test_issue3962.py:
```diff
@@ -0,0 +1,42 @@
+    assert doc2[2].dep_ == "pobj"
+    assert doc2[2].head.text == "at"
+    assert doc2[3].dep_ == "punct"
+    assert doc2[3].head.text == ","      # head set back to itself
```
@svlandeg Sorry for the nitpicky comment, but just for consistency: Can you set your editor to automatically insert newlines at the end of files? Or I think if you set up [`black`](https://github.com/python/black) to auto-format on save, it should also do it by default (at least for Python).

svlandeg commented:
@honnibal : sounds like a good idea and I can definitely implement this. Good idea to use the generic `dep`.

Note however that we'll never recover from this specific example, with the sentence `He jests at scars, that never felt a wound.` where the `ROOT` is `felt` which is not in the span, and there's no common ancestors either, so you'll always end up with two "sentences", unless you introduce some kind of new artificial root (?)

svlandeg commented:
Actually, I'm not sure your third point (fall-back) makes sense: if there's no ancestor in the span, attaching it to the previous ROOT would effectively attach it to the previous sentence?

svlandeg commented:
Ok, so the latest push changes the behaviour to locate the first ancestor within the span. I moved the code to `span.pyx` instead of `doc.pyx` because we need access to the data outside of the span (i.e. the original array) to navigate the ancestors. I also changed the link text to `dep` when the head was changed either to the ancestor, or to itself (in case no ancestor was found).

I added a bit to the unit test that tests a case where this procedure does end up in 1 sentence instead of 2 (when taking `doc[6:9]`, "a" is attached to "felt" instead of to "wound"), though the first example (`doc[1:5]`) still remains two sentences as I explained above.

honnibal commented:
I'm mostly thinking of something like the heuristics in `span.root`: https://github.com/explosion/spaCy/blob/master/spacy/tokens/span.pyx#L490

I wonder whether we could keep it simple and use exactly this? For all words that point outside the span, attach them to the `span.root` with label `dep`? I think that actually does roughly what I described, but would be really easy to implement, and definitely easy to document.

svlandeg commented:
Ok, so right now the implementation goes as such, if a token's `head` is outside the span:
1. Navigate its ancestors by looping through `orig_array` (L256 in `Span`), take the first good ancestor, and set the `DEP` to `dep`  --> we can remove this step again if you like, but I think this keeps the tree as intact as possible

2. If still no suitable ancestor is found, set to the `span.root.i` and set the `DEP` to `dep`

This fixes the first unit test as `,` is now attached to `jests` (after `felt` was found to be out-of-span) and now we have 1 sentence there. Hooray :-)

However - I still think this may artificially merge sentences. I added a new example, taking span [1:7] in the sentence "He jests at scars. They never felt a wound." with `ROOT` tokens `jests` and `felt` respectively. After breaking the doc into the span "jests at scars. They never", both `They` and `never` will get attached to `jests` in the previous sentence. The final span will be one sentence (instead of, arguably, two).

honnibal commented:
I think you're right that we'd want two sentences in that example. Do you see an easy way to improve the logic.

svlandeg commented:
Perhaps we could introduce "artificial new heads", being the first token encountered in a sentence that got "cut up". Basically as soon as we encounter a token of which the root ancestor is outside of the span (and no other ancestors were appropriate), we link that token to itself, and keep track that this token replaces the old root ancestor. Then whenever we encounter another token with the same root (i.e. within the same sentence) we link it to the new artificial one. This should ensure that we'll keep the same number of sentences. What do you think?

svlandeg commented:
Ok, I implemented the above and it seems to work, let me know whether you think this is OK.

Then when testing, I stumbled on this: In the last unit test, the original text is `He jests at scars. They never felt a wound.` When taking span `[1:7]` which should be `jests at scars. They never` and then printing the sentences, the first sentence becomes `jests at scars .` cf [here](https://github.com/explosion/spaCy/blob/3949b4661eea2722dc0d8c8b11f891d09809676f/spacy/tests/regression/test_issue3962.py#L111), note the space before the dot. Is this ok, or does this point to some (other) trouble regarding the copying of the whitespaces?
